worker/src/RTC/TransportCongestionControlServer.cpp

```diff
@@ -109,6 +112,162 @@ namespace RTC
 		return this->packetLoss;
 	}
 
+	void TransportCongestionControlServer::OnPacketArrival(uint16_t sequence_number, uint64_t arrival_time)
+	{
+		if (arrival_time > kMaxTimeMs)
+		{
+			MS_WARN_DEV("arrival time out of bounds:%" PRIu64 "", arrival_time);
+
+			return;
+		}
+
+		int64_t seq = this->unwrapper.Unwrap(sequence_number);
+
+		if (seq <= 0)
+		{
+			MS_WARN_DEV("invalid seq_num:%" PRIu16 ", unwrap-seq:%" PRId64 "", sequence_number, seq);
+
+			return;
+		}
+
+		if (
+		  this->periodicWindowStartSeq &&
+		  this->packetArrivalTimes.lower_bound(*this->periodicWindowStartSeq) ==
+		    this->packetArrivalTimes.end())
+		{
+			// Start new feedback packet, cull old packets.
+			for (auto it = this->packetArrivalTimes.begin(); it != this->packetArrivalTimes.end() &&
+			                                                 it->first < seq &&
+			                                                 arrival_time - it->second >= 500;)
+			{
+				it = this->packetArrivalTimes.erase(it);
+			}
+		}
+
+		if (!this->periodicWindowStartSeq || seq < *this->periodicWindowStartSeq)
+		{
+			this->periodicWindowStartSeq = seq;
+		}
+
+		// We are only interested in the first time a packet is received.
+		if (this->packetArrivalTimes.find(seq) != this->packetArrivalTimes.end())
+		{
+			return;
+		}
+
+		this->packetArrivalTimes[seq] = arrival_time;
+		// Limit the range of sequence numbers to send feedback for.
+		auto first_arrival_time_to_keep = this->packetArrivalTimes.lower_bound(
+		  this->packetArrivalTimes.rbegin()->first - kMaxNumberOfPackets);
+
+		if (first_arrival_time_to_keep != this->packetArrivalTimes.begin())
+		{
+			this->packetArrivalTimes.erase(this->packetArrivalTimes.begin(), first_arrival_time_to_keep);
+
+			// |this->packetArrivalTimes| cannot be empty since we just added one element
+			// and the last element is not deleted.
+			// RTC_DCHECK(!this->packetArrivalTimes.empty());
```

Comment:
Let's not dirty mediasoup own C++ code with libwebrtc stuff. Use `MS_ASSERT()` exported by `Logger.hpp` in mediasoup, as used in tons of files.